accepted/0000-workspace-dependency-sharing.md

```diff
@@ -0,0 +1,283 @@
+# Change which dependencies are shared among workspace projects
+
+## Summary
+
+A common use case for workspaces is to develop multiple related packages as
+a single overarching project, so that they can be more easily kept in sync
+and tested with one another.
+
+The current implementation is somewhat naive.  Workspaces are treated
+roughly the same as `file:` dependencies from the root project, with some
+additional semantics applied.  Any shared dependencies are deduplicated up
+to the top-most `node_modules` folder if possible, as they would be if
+they were simple `file:` dependencies.
+
+This RFC proposes more specific guidelines for which types of dependencies
+are shared in various scenarios.
+
+Implementation details are out of scope for this RFC.  This proposal
+focuses exclusively on the user experience and expectations for which
+packages are shared between workspaces in a project.
+
+## Motivation
+
+There are use cases where the naive workspaces implementation, while
+simple, does not provide the ideal user experience.
+
+### Shadow Dependencies
+
+A naive implementation of workspaces makes it easy to come to rely on
+undeclared dependencies.  For example, consider a project with this
+`package.json` file:
+
+```json
+{
+  "workspaces": [
+    "packages/*"
+  ]
+}
+```
+
+and this folder structure:
+
+```
+root
++-- packages
+    +-- a
+    |   +-- package.json
+    |   +-- index.js
+    +-- b
+        +-- package.json
+        +-- index.js
+```
+
+If `a` declares a dependency on `foo`, and `b` does _not_ declare that
+dependency, then the code in `b` can call `require('foo')` and it will work
+without any errors in development.  But when `b` is published, and
+subsequently installed as a standalone project, it fails to find the `foo`
+dependency.
```

Comment:
we may want to explicitly declare it as a non-goal to address "if `foo` depends on `c`, `a` can `require('c')` even though it doesn't declare that dependency"?

Reply:
Yeah, I think that's covered more directly by `--mode=isolated`.